### ISSUE TITLE
v0.5 release goals

### ISSUE THREAD

**Issue by irakliyk** [facebook/winterfell#94]
For v0.5 release I'm thinking of adding the following to the library:

- [ ] #93 
- [ ] #80 
- [ ] Enable custom logic for constraint evaluation in the prover.
- [x] #52 
- [ ] Add full support for traces with more than 2 segments.

Suggestions/comments are welcome.

**Comment by maxgillett** [comment]
It could be nice to also lift the restriction on the number of auxiliary segments in this release. The required changes are in these commits: https://github.com/maxgillett/winterfell/commit/4d5490f59a5c5e98aa598297765cf4c981c6640e, https://github.com/maxgillett/winterfell/commit/f48ac44b6d2ecb9266f03928cd1aea0741709db5, and https://github.com/maxgillett/winterfell/commit/463ee452278ce78ab29b7708ddff8354a5c73e70, but a test case for the library is still missing.

For the third point do you mean custom divisors? I've implemented constraint-specific transition exemptions in https://github.com/maxgillett/winterfell/commit/6169c6d6abb031f9effc052fab8b55734b56d9b6, but it might be better for performance to make this segment-specific if there isn't a need to specify exemptions at a more granular level. There is some discussion of it in https://github.com/novifinancial/winterfell/pull/91#discussion_r874544533.

**Comment by irakliyk** [comment]
> It could be nice to also lift the restriction on the number of auxiliary segments in this release.

Added!

> For the third point do you mean custom divisors?

I actually meant something different: right now constraint evaluation is shared between prover and verifier (the logic is provide via `evaluate_transition()` functions etc. from the `Air` trait). This is nice because we have to write constraint evaluation logic only once and both the prover and the verifier can use it. But it also limits optimization opportunities in the prover where we know we are working with base field elements and could apply various optimizations based on that.

So, I was thinking of adding something like `evaluate_constraints()` function to the `Prover` trait. The default implementation of this function would cover [these lines]( https://github.com/novifinancial/winterfell/blob/main/prover/src/lib.rs#L287-L303) but then the implementers would be free to override this implementation with something different.

**Comment by cryptopunkchina** [comment]
Hi， I wonder how the  the ARK1 and ARK2 generated for rp64-256 come from? I run the algorithm 5 from <https://eprint.iacr.org/2020/1143.pdf> can not generate the same ARK1 and ARK2

**Comment by matthiasgoergens** [comment]
When is distributed proving coming?

**Comment by irakliyk** [comment]
Superseded by #154
